### PR TITLE
CI: move last Azure job to GHA

### PR DESCRIPTION
**Pull request by andyfaff**

### CODE REVIEW

**Comment by andyfaff** [comment]
The remaining test fails are due to problems with tests, I think.

**Comment by andyfaff** [comment]
@mdhaber @steppi @WarrenWeckesser there's two failures in the BreitWigner distribution tests for the Linux 32-bit job.

Would it be possible to look into those test fails?

**Comment by WarrenWeckesser** [comment]
@andyfaff, I'm taking a look.

One of the tests that fails is a test of round-tripping the distribution through a pickle and unpickle. It is a generic test that is applied to all the distributions.  It is basically this, where `distfn` is the `rel_breitwigner` instance:

```
unpickle = pickle.loads(pickle.dumps(distfn))
args = (36.545206797050334,)
x1 = distfn.cdf(distfn.ppf(0.5, *args), *args)
x2 = unpickle.cdf(unpickle.ppf(0.5, *args), *args)
assert_equal(x1, x2)
```
The values `distfn.ppf(0.5, *args))` and `unpickle.ppf(0.5, *args)`, stored in `medians` in the test, appear to be the same--the error log shows `medians    = [36.517932117503165, 36.517932117503165]`.  The results of the `cdf` calculations, however, differ by 1 ULP.

This appears to be another case where checking floating point results for equality is too optimistic.  Of course, in theory, the unpickled distribution should be doing the exact same calculation with exactly the same input, so that equality comparison looks perfectly fine. I don't know what's going on with that one platform that results in the 1 ULP difference.

**Comment by rgommers** [comment]
Capturing the failures for posterity/searchability:
```
FAILED scipy/stats/tests/test_continuous_basic.py::test_cont_basic[500-200-rel_breitwigner-arg90]
FAILED scipy/stats/tests/test_continuous_basic.py::test_rvs_broadcast[rel_breitwigner-shape_args91]
```

Full output:

<details>

```
________________ test_cont_basic[500-200-rel_breitwigner-arg90] ________________
scipy/stats/tests/test_continuous_basic.py:199: in test_cont_basic
    check_pickling(distfn, arg)
        alpha      = 0.01
        arg        = (36.545206797050334,)
        distfn     = <scipy.stats._continuous_distns.rel_breitwigner_gen object at 0xeb869b38>
        distname   = 'rel_breitwigner'
        locscale_defaults = (0, 1)
        m          = 36.23714553411454
        meths      = [<bound method rv_continuous.pdf of <scipy.stats._continuous_distns.rel_breitwigner_gen object at 0xeb869b38>>, <bound...b38>>, <bound method rv_continuous.logsf of <scipy.stats._continuous_distns.rel_breitwigner_gen object at 0xeb869b38>>]
        n_fit_samples = 200
        rng        = RandomState(MT19937) at 0xE1369568
        rvs        = array([32.94888379, 35.78911124, 35.51452916, 38.61065928, 35.98542226,
       36.73968095, 37.1808344 , 36.80156677, ...84807 , 38.78149111, 36.47484465, 36.37553715,
       35.93138403, 36.82226426, 37.9164441 , 36.20432787, 37.51955919])
        sn         = 500
        spec_x     = {'levy_l': -0.5, 'pareto': 1.5, 'rv_histogram_instance': 5.0, 'truncpareto': 3.2, ...}
        v          = 22.921329819350603
        x          = 0.5
scipy/stats/tests/common_tests.py:292: in check_pickling
    npt.assert_equal(distfn.cdf(medians[0], *args),
E   AssertionError: 
E   Items are not equal:
E    ACTUAL: 0.5000000000000006
E    DESIRED: 0.5000000000000004
        args       = (36.545206797050334,)
        distfn     = <scipy.stats._continuous_distns.rel_breitwigner_gen object at 0xeb869b38>
        medians    = [36.517932117503165, 36.517932117503165]
        r0         = array([39.47342059, 37.6113265 , 36.27151342, 36.5194992 , 36.83279268,
       36.89767328, 36.29633274, 36.61449803])
        r1         = array([39.47342059, 37.6113265 , 36.27151342, 36.5194992 , 36.83279268,
       36.89767328, 36.29633274, 36.61449803])
        rndm       = RandomState(MT19937) at 0xF2FA7E68
        s          = b'\x80\x04\x95\x98&\x00\x00\x00\x00\x00\x00\x8c\x1escipy.stats._continuous_distns\x94\x8c\x13rel_breitwigner_gen\x94\x...>> ax.set_xlim([x[0], x[-1]])\n    >>> ax.legend(loc=\'best\', frameon=False)\n    >>> plt.show()\n    \n\n    \x94ub.'
        unpickled  = <scipy.stats._continuous_distns.rel_breitwigner_gen object at 0xcc049550>
_______________ test_rvs_broadcast[rel_breitwigner-shape_args91] _______________
scipy/stats/tests/test_continuous_basic.py:371: in test_rvs_broadcast
    check_rvs_broadcast(distfunc, dist, allargs, bshape, shape_only, 'd')
        allargs    = [array([[[36.5452068]],

       [[36.5452068]],

       [[36.5452068]],

       [[36.5452068]]]), array([0., 0.]), array([[1.],
       [1.],
       [1.]])]
        bshape     = [4, 3, 2]
        dist       = 'rel_breitwigner'
        distfunc   = <scipy.stats._continuous_distns.rel_breitwigner_gen object at 0xeb869b38>
        k          = 0
        loc        = array([0., 0.])
        nargs      = 1
        scale      = array([[1.],
       [1.],
       [1.]])
        shape_args = (36.545206797050334,)
        shape_only = False
        shp        = (4, 1, 1)
scipy/stats/tests/common_tests.py:336: in check_rvs_broadcast
    assert_allclose(sample, expected, rtol=1e-13)
E   AssertionError: 
E   Not equal to tolerance rtol=1e-13, atol=0
E   
E   Mismatched elements: 24 / 24 (100%)
E   Max absolute difference: 23.65674542
E   Max relative difference: 1.82412254
E    x: array([[[37.674122, 36.285746],
E           [36.704247, 34.061794],
E           [36.293535, 38.482505]],...
E    y: array([[[36.815096, 36.669   ],
E           [36.570289, 31.883147],
E           [36.614877, 36.211121]],...
        allargs    = [array([[[36.5452068]],

       [[36.5452068]],

       [[36.5452068]],

       [[36.5452068]]]), array([0., 0.]), array([[1.],
       [1.],
       [1.]])]
        distfunc   = <scipy.stats._continuous_distns.rel_breitwigner_gen object at 0xeb869b38>
        distname   = 'rel_breitwigner'
        expected   = array([[[36.81509553, 36.66899991],
        [36.57028892, 31.88314734],
        [36.61487724, 36.2111208 ]],

       [...55632]],

       [[36.72088154, 36.48324468],
        [35.7550912 , 36.31952238],
        [32.90234615, 36.44054328]]])
        otype      = 'd'
        rvs        = <numpy.vectorize object at 0xcbf23eb0>
        sample     = array([[[37.67412197, 36.28574625],
        [36.70424743, 34.06179353],
        [36.29353472, 38.48250548]],

       [...76021]],

       [[38.28537953, 36.4248583 ],
        [38.00215128, 33.31073478],
        [35.68992185, 32.32872107]]])
        shape      = [4, 3, 2]
        shape_only = False
=========================== short test summary info ============================
FAILED scipy/stats/tests/test_continuous_basic.py::test_cont_basic[500-200-rel_breitwigner-arg90]
FAILED scipy/stats/tests/test_continuous_basic.py::test_rvs_broadcast[rel_breitwigner-shape_args91]
```

</details>

> I don't know what's going on with that one platform that results in the 1 ULP difference in the result for the `rel_breitwigner` distribution. If no one else sees anything that might cause a difference, then the fix is to change the test to use `assert_allclose` with a very small relative tolerance.

I don't see anything special that'd be pickling-related.

**Comment by andyfaff** [comment]
What's weird is that in the [previous test](kstop) different failures were experienced, these Breit-Wigner fails didn't manifest. I tried running the tests again for the last commit but nothing changes.

The only difference between this new job and the removed Azure job is that here we're running on PyPA supplied images (Ubuntu, as used by cibuildwheel) vs `i386/ubuntu:focal`.

**Comment by rgommers** [comment]
@andyfaff the "previous test" link above is broken, and I can't see what it's meant to be pointing to.

**Comment by andyfaff** [comment]
https://github.com/scipy/scipy/actions/runs/4870583098/jobs/8686512629#step:3:3195

**Comment by rgommers** [comment]
Ah okay, the ones for which there are already tolerance bumps in this PR:
```
FAILED scipy/spatial/transform/tests/test_rotation_spline.py::test_spline_properties
FAILED scipy/special/tests/test_hyp2f1.py::TestHyp2f1::test_region5[hyp2f1_test_case46]
```

<details>

```
 ____________________________ test_spline_properties ____________________________
scipy/spatial/transform/tests/test_rotation_spline.py:128: in test_spline_properties
    assert_allclose(rv0, 0.5 * (rvp + rvm), rtol=1e-15)
E   AssertionError: 
E   Not equal to tolerance rtol=1e-15, atol=0
E   
E   Mismatched elements: 1 / 12 (8.33%)
E   Max absolute difference: 1.11022302e-16
E   Max relative difference: 1.13172884e-15
E    x: array([[-0.12656 ,  0.150715,  0.477349],
E          [ 0.021459,  0.101363,  0.66036 ],
E          [-0.243673,  0.125629,  0.4519  ],
E          [-0.291857,  0.32135 ,  0.234661]])
E    y: array([[-0.12656 ,  0.150715,  0.477349],
E          [ 0.021459,  0.101363,  0.66036 ],
E          [-0.243673,  0.125629,  0.4519  ],
E          [-0.291857,  0.32135 ,  0.234661]])
        angles     = [[-5, 10, 27], [3, 5, 38], [-12, 10, 25], [-15, 20, 11]]
        h          = 1e-08
        rotations  = <scipy.spatial.transform._rotation.Rotation object at 0xd052ce30>
        rv0        = array([[-0.12656021,  0.15071452,  0.47734947],
       [ 0.02145932,  0.10136272,  0.66035993],
       [-0.2436726 ,  0.12562861,  0.45190029],
       [-0.29185722,  0.32134995,  0.23466053]])
        rvm        = array([[-0.12656021,  0.15071452,  0.47734947],
       [ 0.02145932,  0.10136272,  0.66035993],
       [-0.2436726 ,  0.12562861,  0.45190029],
       [-0.29185722,  0.32134995,  0.23466053]])
        rvp        = array([[-0.12656021,  0.15071452,  0.47734947],
       [ 0.02145932,  0.10136272,  0.66035993],
       [-0.2436726 ,  0.12562861,  0.45190029],
       [-0.29185722,  0.32134995,  0.23466053]])
        spline     = <scipy.spatial.transform._rotation_spline.RotationSpline object at 0xcaed9028>
        times      = array([ 0,  5, 15, 27])
 _________________ TestHyp2f1.test_region5[hyp2f1_test_case46] __________________
scipy/special/tests/test_hyp2f1.py:1891: in test_region5
    assert_allclose(hyp2f1(a, b, c, z), expected, rtol=rtol)
E   AssertionError: 
E   Not equal to tolerance rtol=1e-15, atol=0
E   
E   Mismatched elements: 1 / 1 (100%)
E   Max absolute difference: 2.78221093e-15
E   Max relative difference: 1.01034889e-15
E    x: array(0.552767+2.697663j)
E    y: array(0.552767+2.697663j)
        a          = -1.5
        b          = -3.5
        c          = 2.050308316530781
        expected   = (0.5527670397711952+2.697662715303637j)
        hyp2f1_test_case = Hyp2f1TestCase(a=-1.5, b=-3.5, c=2.050308316530781, z=(0.03793103448275881+1.024137931034483j), expected=(0.5527670397711952+2.697662715303637j), rtol=1e-15)
        rtol       = 1e-15
        self       = <scipy.special.tests.test_hyp2f1.TestHyp2f1 object at 0xe4024550>
        z          = (0.03793103448275881+1.024137931034483j)
=========================== short test summary info ============================
FAILED scipy/spatial/transform/tests/test_rotation_spline.py::test_spline_properties
FAILED scipy/special/tests/test_hyp2f1.py::TestHyp2f1::test_region5[hyp2f1_test_case46]
```

</details>

That is a pickling test fails like this, it will be in a flaky matter, is not surprising I think.

**Comment by andyfaff** [comment]
I re-ran the test twice and the fail occurred both times, so I'm not sure whether we skip for 32 bit or relax the tolerance.

**Comment by rgommers** [comment]
Either way would be fine, but maybe skip for 32-bit? It's a generic test, so I'd be curious if it at some point starts failing for any other distribution. If so, we can loosen the tolerance then. If not, it's something related to the extreme value handling of this particular distribution.

**Comment by WarrenWeckesser** [comment]
I don't know what is happening with the `test_rvs_broadcast` test when applied to `rel_breitwigner`.   The following is an explanation of what the test does, but at the moment I don't have a recommendation for fixing or working around the issue. (We *could* add `rel_breitwigner` to the list of distributions in `test_rvs_broadcast` that are tested for shape only, but that would apply to *all* platforms, and it seems like we have just one flaky platform to deal with.)

*Background*

The test `test_rvs_broadcast` calls `check_rvs_broadcast`, which attempts to verify that the arguments to the `rvs` method are broadcast correctly.  It does this by comparing the result of `rvs` when given array arguments to the result of generating the random variates by wrapping the `rvs` method with `np.vectorize`, which will call the `rvs` method repeatedly with scalars.  For many distributions, these should generate the same random variates when using the same RNG.  For some, this won't work, because the `rvs` method makes multiple calls to the underlying RNG for each variate, and the direct call and the `vectorize`d call end up making those calls to the RNG in different orders, resulting in different variates.  For those distributions, only the shape of the result is tested. (In that case, it is a crude test, but at least it will catch new or updated implementations that fail to consider broadcasting.)

`rel_breitwigner` uses the default `rvs` method, which is the inverse transform method. That is, a random variate is generated by applying the `ppf` method to a sample from the uniform distribution U[0,1].  Since only one call to the underlying RNG is required for each variate, the above test should work.  And the tests show that it *does* work on all the other platforms.

`rel_breitwigner` is one of the three distributions that do not provide overrides for `_rvs` and `_ppf`.  The default `ppf` behavior is to numerically invert the CDF.  There are two other distributions that do not override `_rvs` and `_ppf`: `gausshyper` and `studentized_range`.  These two distributions are skipped in `test_rvs_broadcast`, because generating random variates for them is *very* slow.  For what it's worth, if I comment out the code that skips those two distributions, they pass the test locally (64 bit Linux workstation).  But then again, `rel_breitwigner` also passes locally, and it passes on all the other CI platforms, so my local test doesn't really prove much.  It would be interesting to see if `gausshyper` and `studentized_range` pass the test on the troublesome platform.

**Comment by andyfaff** [comment]
Thank you @WarrenWeckesser for looking into that with detail. So do you think we should skip both tests for Linux 32-bit?

**Comment by rgommers** [comment]
The crash for `musllinux` on Cirrus CI looks worrying but is unrelated to this PR. It's in:
```
scipy/optimize/tests/test_nonlin.py", line 320 in test_krylov
```

**Review by rgommers — APPROVED** [review]
This looks great, thanks @andyfaff and @WarrenWeckesser. I propose to squash-merge this. I'm not sure that the `breitwigner` issue is important enough to follow up on, but if so that is best done in a new issue - it should not block this PR.

File: scipy/spatial/transform/tests/test_rotation_spline.py [hunk 1]
```diff
@@ -125,7 +125,8 @@ def test_spline_properties():
     rv0 = spline(times).as_rotvec()
     rvm = spline(times - h).as_rotvec()
     rvp = spline(times + h).as_rotvec()
-    assert_allclose(rv0, 0.5 * (rvp + rvm), rtol=1e-15)
+    # rtol bumped from 1e-15 to 1.5e-15 in gh18414 for linux  bit
```
Comment (WarrenWeckesser) [inline scipy/spatial/transform/tests/test_rotation_spline.py:128]: linux 32 bit?

Comment (andyfaff) [inline scipy/spatial/transform/tests/test_rotation_spline.py:128]:
```suggestion
    # rtol bumped from 1e-15 to 1.5e-15 in gh18414 for linux 32 bit
```

**Comment by andyfaff** [comment]
Feel free to squash-merge away.

**Comment by ilayn** [comment]
Thanks @andyfaff and all!

File: scipy/stats/tests/test_continuous_basic.py [hunk 2]
```diff
@@ -196,7 +198,12 @@ def test_cont_basic(distname, arg, sn, n_fit_samples):
 
     check_named_args(distfn, x, arg, locscale_defaults, meths)
     check_random_state_property(distfn, arg)
-    check_pickling(distfn, arg)
+
+    if distfn in ['rel_breitwigner'] and _IS_32BIT:
```
Comment (WarrenWeckesser) [inline scipy/stats/tests/test_continuous_basic.py:202]: It turns out this is using the wrong variable.  `distfn` is the object; `distname` should be used here.

Comment (andyfaff) [reply scipy/stats/tests/test_continuous_basic.py:202]: Can you make a PR?

Comment (WarrenWeckesser) [reply scipy/stats/tests/test_continuous_basic.py:202]: Yup, I'm just about to click the button on it...

Comment (WarrenWeckesser) [reply scipy/stats/tests/test_continuous_basic.py:202]: https://github.com/scipy/scipy/pull/18493

**Pull request merged**